File: export_player_template/runtime_character_motion.js
// Pure character-stage motion helpers for the exported web runtime.
// Keep normalization and CSS variable shaping here so player.js stays focused on playback.

export const CHARACTER_POSITION_PERCENT = Object.freeze({
  left: 24,
  center: 50,
  right: 76,
});

export const CHARACTER_MOTION_EASING_LABELS = Object.freeze({
  linear: "匀速",
  ease_in: "缓慢起步",
  ease_out: "自然收住",
  ease_in_out: "两端柔和",
  spring: "轻微弹性",
});

export const CHARACTER_MOTION_DURATION_DEFAULT_MS = 600;
export const CHARACTER_MOTION_DURATION_MIN_MS = 0;
export const CHARACTER_MOTION_DURATION_MAX_MS = 10000;

export const DEFAULT_CHARACTER_STAGE = Object.freeze({
  offsetX: 0,
  offsetY: 0,
  scale: 100,
  opacity: 100,
  layer: 0,
  flipX: false,
});

function clamp(value, minimum, maximum) {
  return Math.min(Math.max(value, minimum), maximum);
}

function safeNumber(value, fallback) {
  const number = Number.parseFloat(value ?? "");
  return Number.isFinite(number) ? number : fallback;
}

function safeBoolean(value, fallback = false) {
  if (typeof value === "boolean") return value;
  const normalized = String(value ?? "").trim().toLowerCase();
  if (["true", "1", "yes", "on"].includes(normalized)) return true;
  if (["false", "0", "no", "off"].includes(normalized)) return false;
  return fallback;
}

export function getSafeCharacterPosition(value, fallback = "center") {
  const safeFallback = Object.hasOwn(CHARACTER_POSITION_PERCENT, fallback) ? fallback : "center";
  const position = String(value ?? "").trim();
  return Object.hasOwn(CHARACTER_POSITION_PERCENT, position) ? position : safeFallback;
}

export function getCharacterPositionPercent(value) {
  return CHARACTER_POSITION_PERCENT[getSafeCharacterPosition(value)];
}

export function getPositionOrder(value) {
  return getCharacterPositionPercent(value);
}

export function getSafeCharacterStage(source = {}) {
  const raw = source && typeof source === "object" ? source : {};
  return {
    offsetX: Math.round(clamp(safeNumber(raw.offsetX, DEFAULT_CHARACTER_STAGE.offsetX), -60, 60)),
    offsetY: Math.round(clamp(safeNumber(raw.offsetY, DEFAULT_CHARACTER_STAGE.offsetY), -45, 45)),
    scale: Math.round(clamp(safeNumber(raw.scale, DEFAULT_CHARACTER_STAGE.scale), 45, 220)),
    opacity: Math.round(clamp(safeNumber(raw.opacity, DEFAULT_CHARACTER_STAGE.opacity), 0, 100)),
    layer: Math.round(clamp(safeNumber(raw.layer, DEFAULT_CHARACTER_STAGE.layer), -10, 10)),
    flipX: safeBoolean(raw.flipX, DEFAULT_CHARACTER_STAGE.flipX),
  };
}

export function getCharacterStageFromBlock(block = {}) {
  return getSafeCharacterStage(block.stage ?? block.characterStage ?? {});
}

export function getSafeCharacterMotionEasing(value) {
  const easing = String(value ?? "").trim();
  return Object.hasOwn(CHARACTER_MOTION_EASING_LABELS, easing) ? easing : "ease_out";
}

export function getSafeCharacterMotionDurationMs(value, fallback = CHARACTER_MOTION_DURATION_DEFAULT_MS) {
  const safeFallback = clamp(
    safeNumber(fallback, CHARACTER_MOTION_DURATION_DEFAULT_MS),
    CHARACTER_MOTION_DURATION_MIN_MS,
    CHARACTER_MOTION_DURATION_MAX_MS
  );
  return Math.round(
    clamp(
      safeNumber(value, safeFallback),
      CHARACTER_MOTION_DURATION_MIN_MS,
      CHARACTER_MOTION_DURATION_MAX_MS
    )
  );
}

export function getCharacterMotionCssTimingFunction(value) {
  return {
    linear: "linear",
    ease_in: "cubic-bezier(0.42, 0, 1, 1)",
    ease_out: "cubic-bezier(0.16, 1, 0.3, 1)",
    ease_in_out: "cubic-bezier(0.65, 0, 0.35, 1)",
    spring: "cubic-bezier(0.34, 1.56, 0.64, 1)",
  }[getSafeCharacterMotionEasing(value)];
}

export function buildCharacterMotionEvent(previousState = {}, targetState = {}, block = {}) {
  const previous = previousState && typeof previousState === "object" ? previousState : {};
  const target = targetState && typeof targetState === "object" ? targetState : {};
  return {
    mode: "move",
    characterId: String(target.characterId ?? block.characterId ?? previous.characterId ?? "").trim(),
    previousState: {
      characterId: String(previous.characterId ?? block.characterId ?? "").trim(),
      expressionId: previous.expressionId ?? null,
      expressionName: previous.expressionName ?? "",
      position: getSafeCharacterPosition(previous.position),
      stage: getSafeCharacterStage(previous.stage),
    },
    durationMs: getSafeCharacterMotionDurationMs(block.durationMs),
    easing: getSafeCharacterMotionEasing(block.easing),
  };
}

export function getCharacterStageStyle(stageSource = {}, positionSource = "center") {
  const stage = getSafeCharacterStage(stageSource);
  return [
    `--sprite-position-x:${getCharacterPositionPercent(positionSource)}%;`,
    `--sprite-offset-x:${stage.offsetX}%;`,
    `--sprite-offset-y:${stage.offsetY}%;`,
    `--sprite-scale:${(stage.scale / 100).toFixed(3)};`,
    `--sprite-opacity:${(stage.opacity / 100).toFixed(2)};`,
    `--sprite-layer:${stage.layer};`,
    `--sprite-flip-x:${stage.flipX ? -1 : 1};`,
    `z-index:${20 + stage.layer};`,
  ].join("");
}

export function getCharacterMotionStyle(event = null) {
  if (!event || event.mode !== "move") return "";
  const previous = event.previousState ?? {};
  const stage = getSafeCharacterStage(previous.stage);
  return [
    `--sprite-from-position-x:${getCharacterPositionPercent(previous.position)}%;`,
    `--sprite-from-offset-x:${stage.offsetX}%;`,
    `--sprite-from-offset-y:${stage.offsetY}%;`,
    `--sprite-from-scale:${(stage.scale / 100).toFixed(3)};`,
    `--sprite-from-opacity:${(stage.opacity / 100).toFixed(2)};`,
    `--sprite-motion-ms:${getSafeCharacterMotionDurationMs(event.durationMs)}ms;`,
    `--sprite-motion-easing:${getCharacterMotionCssTimingFunction(event.easing)};`,
  ].join("");
}

File: native_runtime/runtime_character_motion.py
from __future__ import annotations

import math
from typing import Any

try:
    from .runtime_player_view import clamp, get_safe_character_stage
except ImportError:  # pragma: no cover - exported native packages import siblings directly.
    from runtime_player_view import clamp, get_safe_character_stage


CHARACTER_POSITION_RATIOS = {
    "left": 0.24,
    "center": 0.50,
    "right": 0.76,
}
CHARACTER_MOTION_EASINGS = {"linear", "ease_in", "ease_out", "ease_in_out", "spring"}
CHARACTER_MOTION_DURATION_DEFAULT_MS = 600
CHARACTER_MOTION_DURATION_MIN_MS = 0
CHARACTER_MOTION_DURATION_MAX_MS = 10000


def _safe_number(value: Any, fallback: float) -> float:
    try:
        number = float(value)
    except (TypeError, ValueError):
        return float(fallback)
    return number


def get_safe_character_position(value: Any, fallback: str = "center") -> str:
    safe_fallback = fallback if fallback in CHARACTER_POSITION_RATIOS else "center"
    candidate = str(value or "").strip()
    return candidate if candidate in CHARACTER_POSITION_RATIOS else safe_fallback


def get_character_position_ratio(value: Any) -> float:
    return CHARACTER_POSITION_RATIOS[get_safe_character_position(value)]


def get_safe_character_motion_easing(value: Any) -> str:
    candidate = str(value or "").strip()
    return candidate if candidate in CHARACTER_MOTION_EASINGS else "ease_out"


def get_safe_character_motion_duration_ms(
    value: Any,
    fallback: int = CHARACTER_MOTION_DURATION_DEFAULT_MS,
) -> int:
    safe_fallback = clamp(
        _safe_number(fallback, CHARACTER_MOTION_DURATION_DEFAULT_MS),
        CHARACTER_MOTION_DURATION_MIN_MS,
        CHARACTER_MOTION_DURATION_MAX_MS,
    )
    return round(
        clamp(
            _safe_number(value, safe_fallback),
            CHARACTER_MOTION_DURATION_MIN_MS,
            CHARACTER_MOTION_DURATION_MAX_MS,
        )
    )


def apply_character_motion_easing(progress: Any, easing: Any = "ease_out") -> float:
    t = clamp(_safe_number(progress, 0.0), 0.0, 1.0)
    safe_easing = get_safe_character_motion_easing(easing)
    if safe_easing == "linear":
        return t
    if safe_easing == "ease_in":
        return t * t * t
    if safe_easing == "ease_in_out":
        return 4 * t * t * t if t < 0.5 else 1 - pow(-2 * t + 2, 3) / 2
    if safe_easing == "spring":
        # A small bounded overshoot keeps the pose readable while still feeling elastic.
        value = 1 - pow(1 - t, 3) + 0.08 * (1 - t) * math.sin(t * 3.5 * math.pi)
        return clamp(value, 0.0, 1.08)
    return 1 - pow(1 - t, 3)


def build_native_character_motion_state(
    previous_state: dict | None,
    block: dict | None,
    started_at_ms: int,
) -> dict:
    previous = previous_state if isinstance(previous_state, dict) else {}
    source = block if isinstance(block, dict) else {}
    character_id = str(source.get("characterId") or previous.get("characterId") or "").strip()
    previous_position = get_safe_character_position(previous.get("position"))
    target_position = get_safe_character_position(source.get("position"), previous_position)
    previous_stage = get_safe_character_stage(previous.get("stage"))
    target_stage = get_safe_character_stage(source.get("stage") or previous_stage)
    return {
        "characterId": character_id,
        "startedAtMs": int(started_at_ms),
        "durationMs": get_safe_character_motion_duration_ms(source.get("durationMs")),
        "easing": get_safe_character_motion_easing(source.get("easing")),
        "fromState": {
            "characterId": character_id,
            "expressionId": previous.get("expressionId"),
            "position": previous_position,
            "stage": previous_stage,
        },
        "targetState": {
            "characterId": character_id,
            "expressionId": source.get("expressionId") or previous.get("expressionId"),
            "position": target_position,
            "stage": target_stage,
        },
    }


def get_native_character_motion_progress(motion: dict | None, now_ms: int) -> float:
    if not isinstance(motion, dict):
        return 1.0
    duration_ms = get_safe_character_motion_duration_ms(motion.get("durationMs"))
    if duration_ms <= 0:
        return 1.0
    elapsed_ms = max(0, int(now_ms) - int(motion.get("startedAtMs") or 0))
    return clamp(elapsed_ms / duration_ms, 0.0, 1.0)


def is_native_character_motion_complete(motion: dict | None, now_ms: int) -> bool:
    return get_native_character_motion_progress(motion, now_ms) >= 1.0


def _interpolate_number(start: Any, end: Any, progress: float) -> float:
    return _safe_number(start, 0.0) + (_safe_number(end, 0.0) - _safe_number(start, 0.0)) * progress


def get_native_character_render_pose(
    target_state: dict | None,
    motion: dict | None,
    now_ms: int,
) -> dict:
    target = target_state if isinstance(target_state, dict) else {}
    target_stage = get_safe_character_stage(target.get("stage"))
    target_position = get_safe_character_position(target.get("position"))
    if not isinstance(motion, dict):
        return {
            **target,
            "position": target_position,
            "positionRatio": get_character_position_ratio(target_position),
            "stage": target_stage,
        }

    source = motion.get("fromState") if isinstance(motion.get("fromState"), dict) else {}
    source_stage = get_safe_character_stage(source.get("stage"))
    source_position = get_safe_character_position(source.get("position"))
    raw_progress = get_native_character_motion_progress(motion, now_ms)
    progress = apply_character_motion_easing(raw_progress, motion.get("easing"))
    stage = get_safe_character_stage(
        {
            "offsetX": _interpolate_number(source_stage["offsetX"], target_stage["offsetX"], progress),
            "offsetY": _interpolate_number(source_stage["offsetY"], target_stage["offsetY"], progress),
            "scale": _interpolate_number(source_stage["scale"], target_stage["scale"], progress),
            "opacity": _interpolate_number(source_stage["opacity"], target_stage["opacity"], progress),
            "layer": target_stage["layer"],
            "flipX": target_stage["flipX"] if raw_progress >= 0.5 else source_stage["flipX"],
        }
    )
    position_ratio = _interpolate_number(
        get_character_position_ratio(source_position),
        get_character_position_ratio(target_position),
        progress,
    )
    return {
        **target,
        "expressionId": target.get("expressionId") if raw_progress >= 0.5 else source.get("expressionId"),
        "position": target_position,
        "positionRatio": position_ratio,
        "stage": stage,
        "motionProgress": raw_progress,
    }
